tls/s2n_server_hello_retry.c

```diff
@@ -68,5 +74,34 @@ int s2n_server_hello_retry_send(struct s2n_connection *conn)
 
 int s2n_server_hello_retry_recv(struct s2n_connection *conn)
 {
-    S2N_ERROR(S2N_ERR_BAD_MESSAGE);
+    notnull_check(conn);
+
+    /* If the client receives a second HelloRetryRequest in the same connection, 
+    it MUST abort handshake. */
+    if (conn->handshake.client_hrr_received) {
+        S2N_ERROR(S2N_ERR_BAD_MESSAGE);
+    }
+    conn->handshake.client_hrr_received = 1;
```

Comment:
We talked about this offline: each side of the connection switches to a hello retry once, so we can check when we do that switch that we aren't already a hello retry. The flag shouldn't be necessary, just check the handshake type when we go to set the handshake type.